### PR TITLE
PIM-4308: MongoDB indexes are removed on schema update

### PR DESCRIPTION
**Pull request by fitn**
| Q                    | A
| -------------------- | ---
| Bug fix?             | Y
| New feature?         | Y
| BC breaks?           | Y
| CI currently passes? | ?
| Tests pass?          | ?
| Scenarios pass?      | ?
| Checkstyle issues?*  | N (but already broken)
| PMD issues?**        | N (but already broken)
| Changelog updated?   | Y
| Fixed tickets        | PIM-4308
| DB schema updated?   | N
| Migration script?    | N
| Doc PR               | N

### CODE REVIEW

File: src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+use Symfony\Bundle\FrameworkBundle\Command\ContainerAwareCommand;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+/**
+ * Command creating indexes on MongoDB for Product collection
+ *
+ * @author    Romain Monceau <romain@akeneo.com>
+ * @copyright 2014 Akeneo SAS (http://www.akeneo.com)
```
Comment (solivier) [inline src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php]: 2015

File: src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+     * {@inheritdoc}
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $storageDriver = $this->getContainer()->getParameter('pim_catalog_product_storage_driver');
+
+        if ($storageDriver !== 'doctrine/mongodb-odm') {
+            $output->writeln('<error>This command could be only launched on mongodb storage</error>');
+
+            return 1;
```
Comment (solivier) [inline src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php]: -1 ?

File: src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        }
+
+        $indexCreator = $this->getIndexCreator();
+        $indexCreator->ensureUniqueAttributesIndexes();
+        $indexCreator->ensureCompletenessesIndexes();
+        $indexCreator->ensureAttributesIndexes();
+    }
+
+    /**
+     * @return \Pim\Bundle\CatalogBundle\Doctrine\MongoDBODM\IndexCreator
```
Comment (solivier) [inline src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php]: return IndexCreator and use the 'use' statement instead ?

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/IndexCreator.php [hunk 4]
```diff
@@ -131,6 +140,55 @@ public function ensureIndexesFromCurrency(CurrencyInterface $currency)
...
+            self::MONGODB_INDEXES_LIMIT
+        );
+
+        foreach ($attributes as $attribute) {
+            $this->ensureIndexesFromAttribute($attribute);
+        }
+    }
+
+    /**
+     * @return \Pim\Bundle\CatalogBundle\Repository\AttributeRepositoryInterface
```
Comment (solivier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/IndexCreator.php]: return AttributeRepositoryInterface and use the 'use' statement

File: src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        if ($storageDriver !== 'doctrine/mongodb-odm') {
+            $output->writeln('<error>This command could be only launched on mongodb storage</error>');
+
+            return 1;
+        }
+
+        $indexCreator = $this->getIndexCreator();
+        $indexCreator->ensureUniqueAttributesIndexes();
+        $indexCreator->ensureCompletenessesIndexes();
+        $indexCreator->ensureAttributesIndexes();
```
Comment (solivier) [inline src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php]: missing return

File: src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $storageDriver = $this->getContainer()->getParameter('pim_catalog_product_storage_driver');
+
+        if ($storageDriver !== 'doctrine/mongodb-odm') {
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php]: yoda style

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/IndexCreator.php [hunk 7]
```diff
@@ -19,6 +19,9 @@
  */
 class IndexCreator
 {
+    /** @const int */
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/IndexCreator.php]: /** @staticvar int */

**Comment by solivier** [comment]
you broke the specs, you forgot to add the string argument in your specs

**Comment by Nuscly** [comment]
Will it be backported to 1.3 ?
The mongodb update's script should call this command.

**Comment by fitn** [comment]
@Nuscly I hope but I won't do :p

File: src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php (hunk 5 above)
Comment (fitn) [inline src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php]: Done

File: src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $storageDriver = $this->getContainer()->getParameter('pim_catalog_product_storage_driver');
+
+        if ('doctrine/mongodb-odm' !== $storageDriver) {
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php]: AkeneoStorageUtilsExtension::DOCTRINE_MONGODB_ODM

File: src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+ */
+class MongoDBIndexCreatorCommand extends ContainerAwareCommand
+{
+    /**
+     * {@inheritdoc}
+     */
+    protected function configure()
+    {
+        $this
+            ->setName('pim:mongodb:index-fields')
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php]: as it's only related to the catalog and the products I would have name it like the others command.

ie: something like `pim:product:ensure-mongodb-indexes`

what do you think ?

File: src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/IndexCreator.php [hunk 10]
```diff
@@ -19,6 +20,9 @@
  */
 class IndexCreator
 {
+    /** @staticvar int */
+    const MONGODB_INDEXES_LIMIT = 64;
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Doctrine/MongoDBODM/IndexCreator.php]: :+1: possible to add a comment here or in the class phpdoc to explain why it's limited to 64 indexes ? or a link to the doc ?

File: src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php (hunk 9 above)
Comment (fitn) [inline src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php]: @jjanvier In fact, I wanted to override this command in EE to launch on published product but I can create another one too.
As you prefer, it's not a big deal to change ^^

File: src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+            ->setDescription('Create MongoDB indexes');
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php]: Ensure MongoDB indexes of the products (completeness, unique attributes and attributes used in the grid) are created.

File: src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php (hunk 8 above)
Comment (fitn) [inline src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php]: Nice, I found that dirty ^^

File: src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php (hunk 6 above)
Comment (fitn) [inline src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php]: Yoda is back =)

File: src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php (hunk 9 above)
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php]: @fitn EE command are still called `pim:something` (not pimee)

Comment (fitn) [inline src/Pim/Bundle/CatalogBundle/Command/MongoDBIndexCreatorCommand.php]: @jjanvier I mean pim:product should be named pim:published-product, doesn't it?

**Pull request closed without merging**